### ISSUE TITLE
X-Forwarded-Host reports wrong host

### ISSUE THREAD

**Issue by jrudolph** [vulcand/vulcand#269]
It seems X-Forwarded-Host doesn't report the original Host but the host to which the request is forwarded (i.e. the value X-Forwarded-Host and Host are the same) which defeats its purpose.

**Comment by Poogles** [comment]
Have you set TrustForwardHeader to True when upserting the frontend?

**Comment by jrudolph** [comment]
No, I'm not having a proxy in front of vulcan. I'm talking about the headers vulcan generates itself. I want this line to work:

https://github.com/vulcand/oxy/blob/master/forward/rewrite.go#L38

but it seems that at this point `req.Host` doesn't point to the original host any more.

**Comment by Poogles** [comment]
I'm not so familiar with Go so I'll need someone else to confirm, but that line is adding the hostname it's forwarding to as the X-Forwarded-Host due to you not setting TrustFowardHeader to true on line 35.

https://github.com/vulcand/oxy/blob/master/forward/rewrite.go#L35

**Comment by jrudolph** [comment]
As I understand, the question that `TrustFowardHeader` decides is whether *incoming* `X-Forwarded-XYZ` headers should be forwarded to the backend or not. That's not my scenario.

**Comment by jrudolph** [comment]
It seems the undocumented `PassHostHeader` setting somewhat does what I need. Still, I think `X-Forwarded-Host` should be set correctly regardless of that setting.

**Comment by jrudolph** [comment]
I think, in https://github.com/vulcand/oxy/blob/master/forward/fwd.go#L123 the order of execution should be changed, so that the rewriter is executed before the original Host header is overwritten to the new one, so that the Rewriter still has access to the original host and thus, can set it correctly.

**Comment by jrudolph** [comment]
This seems to be a duplicate of https://github.com/vulcand/oxy/issues/18.

**Comment by Poogles** [comment]
Just checked my code and you're correct,

```"Settings": {"TrustForwardHeader": True,
             "PassHostHeader": True}}```

To be honest I think this is prime for a rewrite as it doesn't make much sense at all compared to every other load balancer/web server out there.  This should be the default behaviour, with the option of being able to turn it off?
